src/embeddings/embeddings_calculator.cc

```diff
@@ -138,52 +137,75 @@ class EmbeddingsCalculator : public CalculatorBase {
             if (input_ids_size > max_context_length) {
                 SPDLOG_LOGGER_DEBUG(embeddings_calculator_logger, "Input size {} exceeds max_context_length {}", input_ids_size, max_context_length);
                 return absl::InvalidArgumentError(absl::StrCat("Input length ", input_ids_size, " longer than allowed ", max_context_length));
             }
-
-            embeddingsOutputMap = embeddings_session->infer(embeddingsInputMap);
-            if (embeddingsOutputMap.empty()) {
+            // creating output to avoid copying big tensor
+            ov::Shape outShape = embeddingsInputMap.at(EMBEDDINGS_MODEL_INPUT_IDS_NAME).get_shape();
+            bool foundMatchinEmbeddingOutput{false};
+            std::string outputNameToSet;
+            for (auto& name : embeddings_session->getOutputNames()) {
+                ov::PartialShape outPShape = embeddings_session->getOutputShape(name);
+                if (outPShape.size() != 3)
+                    continue;
+                try {
+                    outShape.emplace_back(outPShape[2].get_length());
+                    foundMatchinEmbeddingOutput = true;
+                    outputNameToSet = name;
+                    break;
+                } catch (std::exception& e) {
+                    LOG(ERROR) << "Failed to get 3rd dimension of output" << outputNameToSet;
+                    return absl::InternalError(absl::StrCat("Failed to get 3rd dimension of output: ", outputNameToSet));
+                }
+            }
+            if (!foundMatchinEmbeddingOutput) {
+                LOG(INFO) << "Failed to find matching output for correct output setting optimization";
+                return absl::InternalError("Could not find output with 3 dimensions in embeddings model");
+            }
+            ov::Tensor outputTensor(embeddings_session->getOutputDatatype(outputNameToSet), outShape);
+            embeddingsInputMap.emplace(std::string(outputNameToSet), std::move(outputTensor));
+            embeddingsOutputMap = std::make_unique<InferenceOutput>(std::move(embeddings_session->infer(embeddingsInputMap)));
+            if (0 == embeddingsOutputMap->size()) {
                 SPDLOG_LOGGER_DEBUG(embeddings_calculator_logger, "No output from embeddings model");
                 return absl::InternalError("No output from embeddings model");
             }
         } catch (const std::exception& e) {
             SPDLOG_LOGGER_DEBUG(embeddings_calculator_logger, "Caught exception from session infer(): {}", e.what());
+            LOG(INFO) << e.what();
             RET_CHECK(false);
         } catch (...) {
             SPDLOG_LOGGER_DEBUG(embeddings_calculator_logger, "Caught unknown exception from session infer()");
             RET_CHECK(false);
         }
 
-        ov::Tensor embeddingsTensor;
-        if (embeddingsOutputMap.size() == 2) {  // GTE
+        const ov::Tensor* embeddingsTensorPtr{nullptr};
+        if (embeddingsOutputMap->size() == 2) {  // GTE
```

Comment:
why do we need this extra condition? We could just search for the input with 3 dims. For models with 1 input it would be found at first time.

Reply:
@dkalinowski

Reply:
That's right, we could do it in both cases